src/System.Net.HttpListener/src/System/Net/Managed/HttpListenerRequest.Managed.cs

```diff
@@ -383,32 +380,46 @@ public bool KeepAlive
 
         public Guid RequestTraceIdentifier => Guid.Empty;
 
-        public IAsyncResult BeginGetClientCertificate(AsyncCallback requestCallback, object state)
+        private IAsyncResult BeginGetClientCertificateCore(AsyncCallback requestCallback, object state)
         {
-            Task<X509Certificate2> getClientCertificate = new Task<X509Certificate2>(() => GetClientCertificate());
-            return TaskToApm.Begin(getClientCertificate, requestCallback, state);
+            var asyncResult = new GetClientCertificateAsyncResult(this, state, requestCallback);
+
+            // The certificate is already retrieved by the time this method is called.
+            ClientCertState = ListenerClientCertState.Completed;
+            asyncResult.InvokeCallback(GetClientCertificate());
```

Comment:
Two problems here:
1. What if GetClientCertificate() throws an exception?  That error is going to be propagated out synchronously from this BeginGet call, but it should really be propagated out as part of the IAsyncResult (I assume that's what happens in the Windows code... it should be).
2. GetClientCertificate is being invoked synchronously, but the purpose of BeginGet is to invoke it asynchronously.  If we have no way of doing that with actual async I/O, the call to GetClientCertificate should be wrapped in either a Task or ThreadPool.QueueUserWorkItem (it was previously wrapped in a Task, but using Task's ctor is incorrect... it should either use Task.Run, Task.Factory.StartNew, or the ctor but with a subsequent call to .Start).